pkg/controller/garbagecollector/healthz.go

```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package garbagecollector
+
+import (
+	"errors"
+	"net/http"
+
+	"k8s.io/apiserver/pkg/server/healthz"
+)
+
+func NewHealthzHandlers(controller *GarbageCollector) []healthz.HealthzChecker {
+	return []healthz.HealthzChecker{&cacheSyncedCheck{controller}}
+}
+
+type cacheSyncedCheck struct {
+	controller *GarbageCollector
+}
+
+func (c cacheSyncedCheck) Name() string {
+	return "cachesSynced"
+}
+
+func (c *cacheSyncedCheck) Check(r *http.Request) error {
```

Comment:
an example that checks if the caches were synced.

running `curl ".../healthz/controllers/garbagecollector?verbose=true"` gives 
```
[+]cachesSynced ok
healthz check passed
```

running `curl ".../healthz/controllers/garbagecollector/cachesSynced"` returns 
```
ok
```

Reply:
This is a great detector, but I think it's a debug handler (first arg).  If we returned unhealthy until our caches sync, then the standby process will constantly crashloop.

Reply:
+1 Davids point. If we believe that the GC has a problem when its cache is too out of date, then we need to define what too out of date is and health check based on being outside that duration.

Reply:
ah, I see, you're absolutely right. 

would it be okay to write a generic detector that would accept multiple `InformerSynced` functions, a `wait` time after which we start to fail  and perhaps optional `resync` time for things like `gc` that needs periodic synchronisation?

Reply:
> would it be okay to write a generic detector that would accept multiple `InformerSynced` functions, a `wait` time after which we start to fail and perhaps optional `resync` time for things like `gc` that needs periodic synchronisation?

I think the informer should try to summarize its health via a function that conforms to a healthcheck.  To do that, it should have limits for number of consistent failures, minimum failures, and minimum time to report unhealthy.

Reply:
drive by comment: you can look at the way workqueue exposes metrics for a model here.